text/0000-exception-safe-event-handlers.md

```diff
@@ -0,0 +1,164 @@
+
+## Low-level API
+
+For cases where you need more control, Neon offers a lower-level primitive, the `schedule_with` method. This method also takes an event launcher, but Neon does not automatically call the event handler. Instead, the event launcher receives a `neon::Context`, the `this` context, and the event handler function object, and is given total control over whether and how to call the event handler.
+
+If the event launcher throws a JavaScript exception, it behaves like an uncaught exception in the Node event loop.
+
+Example for providing the current progress of a background operation:
+
+```rust
+    let mut this = cx.this();
+    let cb = cx.argument::<JsFunction>(0)?;
+    let handler = EventHandler::new(cb);
+    // or:      = EventHandler::bind(this, cb);
+    thread::spawn(move || {
+        for i in 0..100 {
+            // do some work ....
+            thread::sleep(Duration::from_millis(40));
+            // schedule a call into javascript
+            handler.schedule_with(move |cx, this, cb| {
+                // call the event handler callback
+                let args = vec![
+                    cx.null().upcast(),
+                    cx.number(i).upcast()
+                ];
+                cb.call(args)?;
+                Ok(())
+            }
+        }
+    });
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+```rust
+struct EventHandler {
+
+    pub fn new<'a, C: Context<'a>, T: Value>(cx: &C, this: Handle<T>, callback: Handle<JsFunction>) -> Self;
+
+    pub fn schedule<T, F>(&self, arg_cb: F)
+        where T: Value,
+              F: for<'a> FnOnce(&mut EventContext<'a>) -> JsResult<'a, T>,
+              F: Send + 'static;
+
+    pub fn schedule_with<F>(&self, arg_cb: F)
+        where F: FnOnce(&mut EventContext, Handle<JsValue>, Handle<JsFunction>) -> NeonResult<()>,
+              F: Send + 'static;
+
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The type signatures are perhaps a bit more complex. But we should lean on the API docs and examples as the way to explain the API, not the type signatures.
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+Currently the `neon::event` module is protected by a feature flag, and can't be made generally available until we resolve the problem that it has no way to handle JavaScript exceptions.
+
+We should separately propose a `try_catch()` API for wrapping computations that might throw JavaScript exceptions in a closure and converting the result of the computation in a `Result`. We could leave the `neon::event` API as-is and just tell people to use that. But the high-level API would be less ergonomic, since you'd commonly have to wrap everything in a `try_catch`. Also it seems better to decouple the two APIs and release `neon::event` without having to design and implement a complete `try_catch` solution.
```

Comment:
I think there would be value in implementing this API _using_ `try_catch` instead of in C++, but, that is orthogonal to this RFC.

Reply:
I'll add a comment about this to this section.